### PR TITLE
boot: track model's sign key ID, prepare infra for tracking candidate model

### PR DESCRIPTION
**Pull request by bboozzoo**
A preparatory PR for further remodel work. Improve tracking of current model in the modeenv by adding all properties that are measured by secboot (model, brand, sign-key-id, and implicit series which is fixed now). Add fields that will describe the candidate model.

### CODE REVIEW

File: boot/bootstate20.go [hunk 1]
```diff
@@ -769,3 +769,32 @@ func (brs20 *bootState20RecoverySystem) markSuccessful(update bootStateUpdate) (
...
+
+func (brs20 *bootState20Model) markSuccessful(update bootStateUpdate) (bootStateUpdate, error) {
+	u20, err := toBootStateUpdate20(update)
+	if err != nil {
+		return nil, err
+	}
+
+	// sign key ID was not being populated in earlier versions of snapd, try
+	// to remedy that
+	if u20.writeModeenv.SignKeyID == "" {
```
Comment (pedronis) [inline boot/bootstate20.go:787]: shouldn't this be u20.modeenv ?

File: boot/bootstate20.go [hunk 2]
```diff
@@ -769,3 +769,32 @@ func (brs20 *bootState20RecoverySystem) markSuccessful(update bootStateUpdate) (
...
+		newM, err := u20.writeModeenv.Copy()
```
Comment (pedronis) [inline boot/bootstate20.go:788]: why the copy? writeModeenv is already a copy itself. Afaict the rest of code just writes into writeModeenv

**Review by pedronis — COMMENTED** [review]
some questions

File: boot/bootstate20.go (hunk 2 above)
Comment (bboozzoo) [reply boot/bootstate20.go]: i've dropped the copy

File: boot/modeenv.go [hunk 3]
```diff
@@ -59,9 +59,18 @@ type Modeenv struct {
...
 	CurrentKernels      []string `key:"current_kernels"`
-	Model               string   `key:"model"`
-	BrandID             string   `key:"model,secondary"`
-	Grade               string   `key:"grade"`
+	// Model, BrandID, Grade, SignKeyID describe the properties of current
+	// device model.
+	Model     string `key:"model"`
+	BrandID   string `key:"model,secondary"`
+	Grade     string `key:"grade"`
+	SignKeyID string `key:"sign_key_id"`
```
Comment (pedronis) [inline boot/modeenv.go:67]: we should probably follow the naming we used in the boot chains and call this:

ModelSignKeyID  model_sign_key_id

File: boot/modeenv.go [hunk 4]
```diff
@@ -59,9 +59,18 @@ type Modeenv struct {
...
+	// TryModel, TryBrandID, TryGrade, TrySignKeyID describe the properties
+	// of the candidate model.
+	TryModel     string `key:"try_model"`
+	TryBrandID   string `key:"try_model,secondary"`
+	TryGrade     string `key:"try_grade"`
+	TrySignKeyID string `key:"try_sign_key_id"`
```
Comment (pedronis) [inline boot/modeenv.go:73]: same

**Review by pedronis — APPROVED** [review]
thanks, please consider comment about the new field name and consistency

File: boot/modeenv.go [hunk 5]
```diff
@@ -273,7 +290,20 @@ func (m *Modeenv) WriteTo(rootdir string) error {
 		}
 		marshalModeenvEntryTo(buf, "model", &modeenvModel{brandID: m.BrandID, model: m.Model})
 	}
+	// TODO: complain when grade or key are unset
 	marshalModeenvEntryTo(buf, "grade", m.Grade)
+	marshalModeenvEntryTo(buf, "sign_key_id", m.SignKeyID)
+	if m.TryModel != "" || m.TryBrandID != "" {
+		if m.Model == "" {
```
Comment (mvo5) [inline boot/modeenv.go:297]: (nitpick) I wonder if we should add a test for this, but maybe overkill

**Review by mvo5 — APPROVED** [review]
Thanks for this!

File: boot/modeenv.go (hunk 5 above)
Comment (bboozzoo) [reply boot/modeenv.go:297]: I can add one in a followup

**Comment by bboozzoo** [comment]
The failures are unrelated. @mvo5 @pedronis can you land it?

**Pull request merged**